test/spanner.ts

```diff
@@ -0,0 +1,1209 @@
+    case 1:
+      return 'One';
+    case 2:
+      return 'Two';
+    case 3:
+      return 'Three';
+    default:
+      throw new Error(`Unknown or unsupported number: ${num}`);
+  }
+}
+
+describe('Spanner with mock server', () => {
+  let sandbox: sinon.SinonSandbox;
+  const selectSql = 'SELECT NUM, NAME FROM NUMBERS';
+  const invalidSql = 'SELECT * FROM FOO';
+  const insertSql = `INSERT INTO NUMBER (NUM, NAME) VALUES (4, 'Four')`;
+  const fooNotFoundErr = Object.assign(new Error('Table FOO not found'), {
+    code: grpc.status.NOT_FOUND,
+  });
+  const server = new grpc.Server();
+  const spannerMock = mock.createMockSpanner(server);
+  mockInstanceAdmin.createMockInstanceAdmin(server);
+  mockDatabaseAdmin.createMockDatabaseAdmin(server);
+  let spanner: Spanner;
+  let instance: Instance;
+  let dbCounter = 1;
+
+  function newTestDatabase(options?: SessionPoolOptions): Database {
+    return instance.database(`database-${dbCounter++}`, options);
+  }
+
+  before(() => {
+    sandbox = sinon.createSandbox();
+    const port = server.bind(
+      '0.0.0.0:0',
+      grpc.ServerCredentials.createInsecure()
+    );
+    server.start();
+    spannerMock.putStatementResult(
+      selectSql,
+      mock.StatementResult.resultSet(mock.createSimpleResultSet())
+    );
+    spannerMock.putStatementResult(
+      invalidSql,
+      mock.StatementResult.error(fooNotFoundErr)
+    );
+    spannerMock.putStatementResult(
+      insertSql,
+      mock.StatementResult.updateCount(1)
+    );
+
+    // TODO(loite): Enable when SPANNER_EMULATOR_HOST is supported.
+    // Set environment variable for SPANNER_EMULATOR_HOST to the mock server.
+    // process.env.SPANNER_EMULATOR_HOST = `localhost:${port}`;
+    spanner = new Spanner({
+      projectId: 'fake-project-id',
+      servicePath: 'localhost',
+      port,
+      sslCreds: grpc.credentials.createInsecure(),
+    });
+    // Gets a reference to a Cloud Spanner instance and database
+    instance = spanner.instance('instance');
+  });
+
+  after(() => {
+    server.tryShutdown(() => {});
+    delete process.env.SPANNER_EMULATOR_HOST;
+    sandbox.restore();
+  });
+
+  describe('basics', () => {
+    it('should return different database instances when the same database is requested twice with different session pool options', async () => {
+      const dbWithDefaultOptions = newTestDatabase();
+      const dbWithWriteSessions = instance.database(dbWithDefaultOptions.id!, {
+        writes: 1.0,
+      });
+      assert.notStrictEqual(dbWithDefaultOptions, dbWithWriteSessions);
+    });
+
+    it('should execute query', async () => {
+      // The query to execute
+      const query = {
+        sql: selectSql,
+      };
+      const database = newTestDatabase();
+      try {
+        const [rows] = await database.run(query);
+        assert.strictEqual(rows.length, 3);
+        let i = 0;
+        rows.forEach(row => {
+          i++;
+          assert.strictEqual(row[0].name, 'NUM');
```

Comment:
optional: you can destructure instead of calling by index

Reply:
Thanks for pointing that out (I'm quite new to Typescript / JavaScript). I tried it, but it turns out that the declared return type of `database.run(...)` is incorrect. It claims to return a `Promise<Row[]>`, but in reality it returns a `Promise<RowResponse>`, where `RowResponse` is a `[Row[], ResultSetStats]`. That prevents me from using destructuring here, as the typing is incorrect. I'll open a separate issue for this and look into the destructuring later for this part of the code.